pkg/kv/kvserver/replica_circuit_breaker.go

```diff
@@ -61,12 +62,141 @@ type replicaCircuitBreaker struct {
 	stopper *stop.Stopper
 	r       replicaInCircuitBreaker
 	st      *cluster.Settings
+	cancels CancelStorage
 	wrapped *circuit.Breaker
+
+	versionIsActive int32 // atomic
+}
+
+// Register takes a cancelable context and its cancel function, and registers
+// them with the circuit breaker. If the breaker is already tripped, its error
+// is returned immediately and the caller should not continue processing the
+// request. Otherwise, the cancel function is invoked when the breaker trips.
```

Comment:
This should mention that the caller must also invoke `cancel()` to avoid a context leak in the happy case. Also, s/when/when and if/.